Contributing OpenAI GPT-3 Models

rguan1 opened the pull request:
**Patch description**

This pull request adds support for GPT-3 models via OpenAI's python [completion endpoint](https://platform.openai.com/docs/api-reference/completions/create). This is a followup on this [issue](https://github.com/facebookresearch/ParlAI/issues/4555) that discussed how to integrate a model. I've followed the steps at the end of the linked issue to help properly implement the model for ParlAI.

**Testing steps**
We have manually tested the changes and confirmed that they work for both `self-chat` and `interactive` commands

```
(chatbot) Singularity> parlai interactive -m gpt3 --openai-api-key <insert your api key> --max-tokens 40 --model-name text-ada-001

Enter [DONE] if you want to end the episode, [EXIT] to quit.
20:33:00 | creating task(s): interactive
Enter Your Message: How are you doing today ada?
[Gpt3Agent]: I'm doing well, what do you mean by "aduate"?
Enter Your Message: I am referring to the fact that your model name is text-ada-001
[Gpt3Agent]: The name of your model is text-ada-001.
Enter Your Message: 
```

```
(chatbot) Singularity> parlai self_chat -m gpt3 --num-self-chats 1 --selfchat-max-turns 2 --openai-api-key <insert your API key> --max-tokens 40 --model-name text-davinci-003 --init-prompt 'You are trying to make a pull request to an open source repo and want to convince the maintainers to let your pull request through on Github'

[context]: Hi!
[Gpt3Agent_1]: Hi there! How can I help?
   [Gpt3Agent_2]: I'd like to make a pull request to an open source repository on Github and would like to convince the maintainers to let my pull request through. Can you provide any guidance on how to do
```

rguan1 commented:
How do I add the openai dependency? I noticed that the build_website ci/cd is failing because the package is not available?

rguan1 commented:
@klshuster Thanks for responding! Got it. Which of the two options is preferable? Also, I am unfamiliar with safe import. How do you do that?

mojtaba-komeili commented:
> @klshuster Thanks for responding! Got it. Which of the two options is preferable? Also, I am unfamiliar with safe import. How do you do that?

For example something like this that we have somewhere in our code:
```.py
try:
    import transformers
except ImportError:
    raise ImportError('Please run `pip install transformers`.')
```

klshuster commented on parlai/agents/gpt3/gpt3.py:
```diff
@@ -0,0 +1,157 @@
+        """
+        self.observation = None
+        self.turns = ''
+
```
nit: extra newline

rguan1 replied:
Fixed!

klshuster reviewed (commented):
i think this looks great! could you merge main into your branch so that tests pass? then i think we can go ahead and merge this

rguan1 commented:
So, it looks like the build_website still wants the package explicitly in requirements.txt. Should I put it in then or is it ok if that test doesn't pass?

Overall, the package seems pretty small https://pypi.org/project/openai/#files . So, adding it shouldn't be a big burden.

rguan1 commented:
I also plan on adding a GPT4/ChatGPT variant which uses a separate API endpoint soon. So, once I figure out how to iron out CI/CD issues in this PR, I'll make one for the other endpoint in the upcoming weeks. So be on the look out for that! ^.^

klshuster commented:
Sure, you can include it, but please pin it to a version (e.g., `openai<=1.0.1`). And, make sure to update the [cache keys in the circle CI config](https://github.com/facebookresearch/ParlAI/blob/86ec7355ed820dea1ecffd04583f3732beb46a78/.circleci/config.yml#L238) if you do so (please simply change the dates on all keys that look like the linked one to the date you make the change)

rguan1 commented:
Added openai to requirements.txt and changed the date to circleci config.yaml. For the yaml changes, I pattern matched against this PR: https://github.com/facebookresearch/ParlAI/commit/104fb967172ff80cac24e22026341cc84253912a

klshuster commented:
Could you pin to a `<=` version instead of `>=`? you can also do `~=` to pin exactly to the version

rguan1 commented:
Hopefully that resolves the last of the lint errors. What linter does the lint ci/cd use? It seems to complain about formatting that the autoformat.sh (black autoformat) is ok with?

mojtaba-komeili commented on requirements.txt:
```diff
@@ -59,3 +59,4 @@ protobuf<=3.20.3, >=3.8.0
 contractions~=0.1.72
 fsspec~=2022.2.0
 google-api-core<=2.11.0
+openai<=0.27.7
```
Now that we have the safe import, there is no need to install this for everyone. Please remove it.

rguan1 replied:
So without adding it to the the requirements.txt, the "build_website" job fails.

I asked about it earlier in the PR: "_So, it looks like the build_website still wants the package explicitly in requirements.txt. Should I put it in then or is it ok if that test doesn't pass?_" @klshuster said that it was ok to add in the followup. 

I don't have an opinion whether or not the package should be added or removed, but I just want to double check on whether or not its important that "build_website" needs to pass before the PR is merged.

mojtaba-komeili replied:
Sorry, I missed part of the previous conversation you had. It makes sense looking back at it.

mojtaba-komeili reviewed (approved):
Thank you for your contribution to our OSS.

rguan1 commented:
Thanks for reviewing this! Do I merge the PR or do y'all (maintainers) merge the PR?

mojtaba-komeili commented:
You can merge it yourself. It always feels good to merge ;)

rguan1 commented:
This is my first time contributing to ParlAI. Normally you'd expect a merge button on a the PR once the PR is approved right? 

This is what I see at the bottom of this PR. I don't see any merge button which suggests to me that the master branch is protected, so I can't merge it? 
<img width="904" alt="image" src="https://github.com/facebookresearch/ParlAI/assets/20273088/8e3d544b-9700-435a-a9e2-44c29fdec19b">

mojtaba-komeili commented:
Oh, OK. I was not aware of that. Merging this. Thanks!
